# Airbnb project

## Introduction

# Introduction

Airbnb has become one of the largest short-term rental platforms in the world, providing hosts with opportunities to generate income while offering travelers a wide range of accommodation options. Because listing prices are influenced by numerous factors—including location, property characteristics, amenities, reviews, and availability—determining an appropriate nightly rate can be challenging.

The goal of this analysis is to explore Airbnb listing data and develop machine learning models capable of predicting nightly prices. Through exploratory data analysis (EDA), we will examine the relationships between listing features and pricing, identify key drivers of price variation, and prepare the data for predictive modeling.

To estimate nightly prices, two regression models will be evaluated: Support Vector Regression (SVR) and Gradient Boosting Regressor. Model performance will be assessed using standard regression metrics, including Root Mean Squared Error (RMSE), Mean Absolute Error (MAE), and the coefficient of determination (R²).

The insights gained from this analysis will serve as the foundation for a broader investment-focused study comparing Airbnb markets across major U.S. cities and evaluating potential revenue opportunities for short-term rental investors.


Dataset was downloaded from kaggle. Link: https://www.kaggle.com/datasets/youssefkhaled117/airbnb
The dataset contains information about listing characteristics, host details, geographic location, pricing structure, availability, and customer feedback.


# Project Goals

* Identify lucrative cities for short-term rentals and estimate potential returns.
* Build a predictive model to estimate the optimal nightly price based on the available data.
* Create a dashboard that highlights the features most strongly associated with booking activity.

This project will primarily use Python, leveraging libraries such as Pandas, Matplotlib, and Seaborn for data analysis and visualization. Tableau will be used to create an interactive dashboard that allows users to explore trends and key performance indicators across Airbnb listings.

## Predictive Pricing Model

Since the target variable, nightly price, is numerical, this problem will be approached using regression techniques. Two regression models will be developed and evaluated:

1. **Support Vector Regression (SVR)** – SVR is capable of capturing complex, non-linear relationships between listing characteristics and pricing, making it well-suited for Airbnb datasets.

2. **Gradient Boosting Regressor** – Gradient Boosting is an ensemble learning method that builds a series of decision trees to improve prediction accuracy. It is particularly effective at modeling non-linear relationships and interactions among features, often achieving strong performance on structured tabular data.

The performance of both models will be evaluated and compared using standard regression metrics such as Root Mean Squared Error (RMSE), Mean Absolute Error (MAE), and the coefficient of determination (R²). The best-performing model will be used to generate pricing recommendations based on listing attributes and market conditions.


# Data Analysis with Python

In [36]:
# Import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [37]:
# Load the dataset
df = pd.read_csv('final_data_with_changes.csv')

# Display the first few rows of the dataset, check missing values, and get a summary of the dataset
df.head()
df.info()
df.describe()

# # Check for missing values
# missing_values = df.isnull().sum()
# print(missing_values)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51707 entries, 0 to 51706
Data columns (total 36 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   price_total                     51707 non-null  float64
 1   room_type                       51707 non-null  object 
 2   is_shared_room                  51707 non-null  int64  
 3   is_private_room                 51707 non-null  int64  
 4   max_guests                      51707 non-null  float64
 5   is_superhost                    51707 non-null  int64  
 6   is_multi_listing                51707 non-null  int64  
 7   is_business_listing             51707 non-null  int64  
 8   cleanliness_score               51707 non-null  float64
 9   guest_satisfaction_score        51707 non-null  float64
 10  num_bedrooms                    51707 non-null  int64  
 11  distance_city_center            51707 non-null  float64
 12  distance_metro                  

,price_total,is_shared_room,is_private_room,max_guests,is_superhost,is_multi_listing,is_business_listing,cleanliness_score,guest_satisfaction_score,num_bedrooms,...,Safety_Index,Monthly_Average_Net_salary,Meal_at_Inexpensive_Restaurant,Taxi_price_per_Km,Monthly_Basic_Utilities,Monthly_Rent_One_Bedroom_CC,Monthly_Rent_One_Bedroom_OCC,Monthly_Rent_Three_Bedroom_CC,Monthly_Rent_Three_Bedroom_OCC,proximity_index
count,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.00000,...,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000,51707.000000
mean,368.620987,0.007078,0.361518,3.161661,0.744097,0.291353,0.350204,9.390624,92.628232,1.15876,...,53.630166,2611.825205,16.499909,1.780522,228.501078,1518.019100,1105.005234,2692.939883,1945.483547,0.597525
std,509.717535,0.083835,0.480445,1.298545,0.436372,0.454390,0.477038,0.954868,8.945531,0.62741,...,10.789932,1179.745274,3.722901,0.737444,73.984517,700.342853,512.655685,1034.869468,830.160945,0.248780
min,7.657682,0.000000,0.000000,2.000000,0.000000,0.000000,0.000000,2.000000,20.000000,0.00000,...,42.000000,1164.490000,10.470000,0.900000,134.930000,614.420000,519.210000,1142.700000,946.390000,0.047455
25%,157.537810,0.000000,0.000000,2.000000,0.000000,0.000000,0.000000,9.000000,90.000000,1.00000,...,44.400000,1410.600000,15.000000,1.150000,184.560000,1280.450000,820.590000,2162.900000,1541.670000,0.432760
50%,260.500210,0.000000,0.000000,3.000000,1.000000,0.000000,0.000000,10.000000,95.000000,1.00000,...,52.700000,2102.820000,15.000000,1.500000,187.370000,1387.500000,1036.000000,2431.820000,1633.330000,0.575571
75%,433.668497,0.000000,1.000000,4.000000,1.000000,1.000000,1.000000,10.000000,99.000000,1.00000,...,66.100000,3305.140000,18.250000,2.800000,291.200000,1398.590000,1061.180000,3088.460000,2145.000000,0.725286
max,43955.618778,1.000000,1.000000,6.000000,1.000000,1.000000,1.000000,10.000000,100.000000,10.00000,...,74.300000,4502.100000,23.010000,3.000000,347.110000,2788.440000,2029.850000,4402.560000,3416.390000,1.820960


# Regression Analysis

We will start with a simple linear regression model, and depending on the result, we will move onto a XGBOOST model which tends to perform better at this types of problems, where multiple features are involved in determining the target variable. 

In [ ]:
# Regression analysis

# Drop features that are not relevant for the regression model
df = df.drop(['longitude', 'latitude', 'country_code', 'country_name', 'attraction_index', 'restaurant_index'], axis=1)

# One hot encode categorical variables if necessary (room_type, city, day_type, district, state, country_code, country_name)
categorical_features = ['room_type', 'city', 'day_type', 'district', 'state']
df = pd.get_dummies(df, columns=categorical_features, drop_first=True)

# Normalize numerical features if necessary (e.g., price_per_night, number_of_reviews, review_scores_rating)
numerical_features = ['cleanliness_score', 'guest_satisfaction_score', 'distance_city_center', 'distance_metro', 'Monthly_Average_Net_salary', 
                     'Monthly_Basic_Utilities', 'Monthly_Rent_One_Bedroom_CC', 'Monthly_Rent_One_Bedroom_OCC', 
                      'Monthly_Rent_Three_Bedroom_CC', 'Monthly_Rent_Three_Bedroom_OCC', 'proximity_index']

df[numerical_features] = (df[numerical_features] - df[numerical_features].mean()) / df[numerical_features].std()

# price_total is the target variable, and the rest are features
X = df.drop('price_total', axis=1)
y = df['price_total']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and fit the linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f'Mean Squared Error: {mse}')
print(f'R^2 Score: {r2}')



Mean Squared Error: 147227.20150863862
R^2 Score: 0.3620221160261855


/Users/julian/Documents/github_repo/Airbnb_project/.venv/lib/python3.9/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/julian/Documents/github_repo/Airbnb_project/.venv/lib/python3.9/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/julian/Documents/github_repo/Airbnb_project/.venv/lib/python3.9/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_


## Model Performance Analysis

### Mean Squared Error (MSE): 147,227.20

The MSE represents the average squared difference between predicted and actual prices. A value of 147,227.20 indicates the model's predictions deviate significantly from actual prices on average. In practical terms, this translates to an average error of approximately **$383 per night** (√147,227 ≈ 383), which is substantial given the price range in the dataset.

### R² Score: 0.362

The R² score of 0.362 means the linear regression model explains only **36.2% of the variance** in nightly prices. This indicates:

- **Poor Fit**: The model captures less than two-thirds of the price variation
- **Missing Factors**: Important pricing drivers may not be adequately captured by the current features
- **Non-linear Relationships**: Linear regression may be too simplistic for this complex dataset

### Interpretation

The combination of high MSE and low R² suggests:

1. **Linear regression is insufficient** for this problem, supporting the project plan to test SVR and Gradient Boosting models
2. **Feature engineering may be needed** to better capture price drivers
3. **Non-linear relationships** likely exist between listing characteristics and pricing that linear models cannot capture
